File: frontend/src/lib/granularity.ts
import type { AccountType } from '@/types'

// Vendor Central reports monthly, Seller Central daily. A view that mixes both
// has to plot each series on its own cadence, so the UI resolves the natural
// granularity of the in-scope accounts and adapts charts and the group-by.
export type Granularity = 'daily' | 'monthly' | 'mixed' | 'unknown'

export function granularityForAccountTypes(
  accountTypes: Iterable<AccountType>
): Granularity {
  let hasSeller = false
  let hasVendor = false
  for (const type of accountTypes) {
    if (type === 'vendor') hasVendor = true
    else if (type === 'seller') hasSeller = true
  }

  if (hasSeller && hasVendor) return 'mixed'
  if (hasVendor) return 'monthly'
  if (hasSeller) return 'daily'
  return 'unknown'
}

interface AccountLike {
  id: string
  account_type: AccountType
}

// Resolve granularity from the full account list and the selected ids. An empty
// selection means "all accounts", matching how the analytics filters behave.
export function granularityForSelection(
  accounts: AccountLike[] | undefined,
  selectedAccountIds: string[]
): Granularity {
  if (!accounts || accounts.length === 0) return 'unknown'
  const inScope =
    selectedAccountIds.length === 0
      ? accounts
      : accounts.filter((account) => selectedAccountIds.includes(account.id))
  return granularityForAccountTypes(inScope.map((account) => account.account_type))
}

File: backend/app/services/granularity.py
"""Resolve the natural time granularity of a set of Amazon accounts.

Vendor Central data is reported monthly (one settled row per month); Seller
Central data is daily. When both are mixed in a single view the series need to
be plotted on different cadences, so callers ask this helper which granularity
applies to the accounts in scope and surface it to the frontend.
"""
from enum import Enum
from typing import Iterable, List, Optional
from uuid import UUID

from sqlalchemy import select

from app.models.amazon_account import AccountType, AmazonAccount


class Granularity(str, Enum):
    DAILY = "daily"
    MONTHLY = "monthly"
    MIXED = "mixed"
    UNKNOWN = "unknown"


def granularity_for_account_types(account_types: Iterable[AccountType]) -> Granularity:
    """Resolve granularity from the account types already in hand."""
    has_seller = False
    has_vendor = False
    for account_type in account_types:
        if account_type == AccountType.VENDOR:
            has_vendor = True
        elif account_type == AccountType.SELLER:
            has_seller = True

    if has_seller and has_vendor:
        return Granularity.MIXED
    if has_vendor:
        return Granularity.MONTHLY
    if has_seller:
        return Granularity.DAILY
    return Granularity.UNKNOWN


async def resolve_granularity(
    db,
    organization_id: UUID,
    account_ids: Optional[List[UUID]] = None,
) -> Granularity:
    """Resolve granularity for the organization's accounts (optionally scoped).

    When ``account_ids`` is empty/None the whole organization is considered,
    matching how the analytics endpoints treat an unscoped request.
    """
    query = select(AmazonAccount.account_type).where(
        AmazonAccount.organization_id == organization_id
    )
    if account_ids:
        query = query.where(AmazonAccount.id.in_(account_ids))

    rows = (await db.execute(query)).scalars().all()
    return granularity_for_account_types(rows)
